### PR TITLE
Director Velocity Fixes

### PR DESCRIPTION
**Pull request by drfiveminusmint**
Currently, director velocity is very inconsistent with sharp vertical angles, due to an oversight in how the target vector is normalised. This PR changes the normalisation to happen after the vertical component has been set to zero.

### CODE REVIEW

**Review by TylerS1066 — COMMENTED** [review]
After re-familiarizing myself with yarro's cryptic variable names, I believe this is mainly correct.

I would like to change `tv` to `tntVector` to improve readability.  In addition, is there some way to simplify the `targetVector.getX() - tv.getX() > 0.7` portions to a min/max call?


I also think there may be a problem with the speed multiplication, since it will take the length of the 3D vector from the TNT, ignore that for the normalization etc, then multiply the version with 0 vertical velocity by that.  Is there a Vector2D we could use instead?  If not, we should set the Y value of `tntVector` to zero before reading the XZ speed.

**Comment by drfiveminusmint** [comment]
interestingly, as intelij helpfully pointed out to me, the conditional ```targetBlock == null``` is always false. This section of code needs some cleanup.

**Comment by drfiveminusmint** [comment]
because of this, directors will always attempt to converge at  a block exactly ```Config.CannonDirectorDistance``` away. This could potentially cause some odd behaviour.

**Comment by TylerS1066** [comment]
> interestingly, as intelij helpfully pointed out to me, the conditional `targetBlock == null` is always false. This section of code needs some cleanup.

I'd be hesitant to remove this.  While spigot's annotation claims it will never be null, I suspect it might in a older version than 1.13.




> because of this, directors will always attempt to converge at a block exactly `Config.CannonDirectorDistance` away. This could potentially cause some odd behaviour.

That argument is a max distance it will search, not an exact bound.   I'm not exactly sure what it returns when nothing is found, so I suspect it is null or the air block at the max distance.

**Comment by drfiveminusmint** [comment]
I'll jury-rig a testing system and test if it can find a null target block on 1.10.2 or 1.12.2

**Comment by drfiveminusmint** [comment]
As I suspected, It does not return null on old versions, rather, it returns an air block at either the max distance or height limit. Unsure about the effect of render distance.

**Comment by drfiveminusmint** [comment]
This is now a much bigger 'fix', however this should make directors work more 'as intended.'

File: src/main/java/net/countercraft/movecraft/combat/movecraftcombat/directors/CannonDirectorManager.java [hunk 1]
```diff
@@ -154,34 +155,27 @@ private void processDirectors() {
...
+                tntVector = tntVector.normalize(); // you normalize it for comparison with the new direction to see if we are trying to steer too far
                 Block targetBlock = p.getTargetBlock(Config.Transparent, Config.CannonDirectorRange);
                 Vector targetVector;
-                if (targetBlock == null) { // the player is looking at nothing, shoot in that general direction
+                // This is horrible but I'm not sure how else to do it, targetBlock == null doesn't work because getTargetBlock never returns null.
+                if (targetBlock.getType().equals(Material.AIR)) { // the player is looking at nothing, shoot in that general direction
                     targetVector = p.getLocation().getDirection();
                 } else { // shoot directly at the block the player is looking at (IE: with convergence)
                     targetVector = targetBlock.getLocation().toVector().subtract(tnt.getLocation().toVector());
-                    targetVector = targetVector.normalize();
```
Comment (drfiveminusmint) [inline src/main/java/net/countercraft/movecraft/combat/movecraftcombat/directors/CannonDirectorManager.java:166]: This vector.normalize call is unnecessary as the vector is normalized a few lines down

**Comment by drfiveminusmint** [comment]
ignore this for a second

**Comment by drfiveminusmint** [comment]
alright this should be ready for review now

File: src/main/java/net/countercraft/movecraft/combat/movecraftcombat/directors/CannonDirectorManager.java [hunk 2]
```diff
@@ -1,6 +1,7 @@
 package net.countercraft.movecraft.combat.movecraftcombat.directors;
 
 import java.util.*;
+import java.util.logging.Level;
```
Comment (TylerS1066) [inline src/main/java/net/countercraft/movecraft/combat/movecraftcombat/directors/CannonDirectorManager.java]: Un-needed import

File: src/main/java/net/countercraft/movecraft/combat/movecraftcombat/directors/CannonDirectorManager.java [hunk 3]
```diff
@@ -154,36 +155,27 @@ private void processDirectors() {
...
                     continue;
                 }
-                Vector tv = tnt.getVelocity();
-                double speed = tv.length(); // store the speed to add it back in later, since all the values we will be using are "normalized", IE: have a speed of 1
-                tv = tv.normalize(); // you normalize it for comparison with the new direction to see if we are trying to steer too far
+                Vector tntVector = tnt.getVelocity();
+                // Store the speed to add it back in later, since all the values we will be using are "normalized", IE: have a speed of 1
+                // We're only interested in the horizontal speed for now since that's all directors *should* affect.
+                double horizontalSpeed = new Vector(tntVector.getX(),0,tntVector.getY()).length();
+                tntVector = tntVector.normalize(); // you normalize it for comparison with the new direction to see if we are trying to steer too far
```
Comment (TylerS1066) [inline src/main/java/net/countercraft/movecraft/combat/movecraftcombat/directors/CannonDirectorManager.java:162]: This normalize still takes into account the vertical vector.  I'd instead create the tntVector using the method you used for horizantalSpeed.

Comment (drfiveminusmint) [reply src/main/java/net/countercraft/movecraft/combat/movecraftcombat/directors/CannonDirectorManager.java:162]: doesn't the `tntVector.setY(0) account for the vertical vector?`

Comment (drfiveminusmint) [reply src/main/java/net/countercraft/movecraft/combat/movecraftcombat/directors/CannonDirectorManager.java:162]: oh, wait, for some reason it seems to still be showing an old version of this file

File: src/main/java/net/countercraft/movecraft/combat/movecraftcombat/directors/CannonDirectorManager.java [hunk 4]
```diff
@@ -154,36 +155,27 @@ private void processDirectors() {
                 if (p == null || p.getInventory().getItemInMainHand().getType() != Config.DirectorTool) {
                     continue;
                 }
-                Vector tv = tnt.getVelocity();
-                double speed = tv.length(); // store the speed to add it back in later, since all the values we will be using are "normalized", IE: have a speed of 1
-                tv = tv.normalize(); // you normalize it for comparison with the new direction to see if we are trying to steer too far
+                Vector tntVector = tnt.getVelocity();
+                // Store the speed to add it back in later, since all the values we will be using are "normalized", IE: have a speed of 1
+                // We're only interested in the horizontal speed for now since that's all directors *should* affect.
+                double horizontalSpeed = new Vector(tntVector.getX(),0,tntVector.getY()).length();
```
Comment (TylerS1066) [inline src/main/java/net/countercraft/movecraft/combat/movecraftcombat/directors/CannonDirectorManager.java]: Uh, I think you meant getX getZ.

File: src/main/java/net/countercraft/movecraft/combat/movecraftcombat/directors/CannonDirectorManager.java (hunk 3 above)
Comment (TylerS1066) [reply src/main/java/net/countercraft/movecraft/combat/movecraftcombat/directors/CannonDirectorManager.java:162]: I don't see a setY anywhere but the end

**Comment by drfiveminusmint** [comment]
closing and re-opening

**Pull request closed without merging**